taxjar/taxjar-magento2-extension
# INT-2041: Command flag option to force API transaction sync

### Context

Users have reached out regarding issues with syncing M2 transactions where the state of the order has changed, and those changes need to be reflected in the TaxJar service, but due to the comparison of a transaction's `updated_at `and `tj_salestax_last_sync` dates, the transaction will continue to be skipped.

Currently, the only acceptable workaround for this problem is for M2 administrators to manually modify date values in DB, which for a large number of transactions.

### Description

- Adds flag `-f|--force` to CLI command `taxjar:transaction:sync` ![image](https://user-images.githubusercontent.com/47947793/137945544-7790ac7c-7d8d-4b05-9fad-ddc69dc47653.png)

- Refactors methods `\Taxjar\SalesTax\Model\Transaction\Order::push()` and `\Taxjar\SalesTax\Model\Transaction\Refund::push()` to accept optional parameters for http method and force-flag option.
- Adds "Force" select element (and tooltip) to the TaxJar transaction sync form 
![image](https://user-images.githubusercontent.com/47947793/138459614-06e1b765-539a-46a5-94c8-bf1ec19f7152.png)

- Adds unit test coverage for `\Taxjar\SalesTax\Model\Transaction\Backfill::process()` with `force` feature

### Performance

This should not have much if any affect on current extension performance,  but should resolve an on-going customer issue.

### Testing

1. With TaxJar's transaction sync feature enabled, *place an order*, either as customer, or as admin on behalf of customer, then invoice and mark shipped as necessary to move order status to "complete".
2. Verify that the order transaction has been persisted to TaxJar.
3. In DB, manually modify a value on the order such as an address value, or the order's `created_at` value.
4. From extension UI, attempt to sync transactions WITHOUT force-flag enabled. (using a date range containing the order)
5. Observe values not updated in TaxJar.
6. From extension UI, attempt to sync transactions WITH force-flag enabled (again, with proper date range).
7. Observe values updated in TaxJar.

#### Versions
